README.md
# Radio Browser

A library to fetch data from the [radio-browser.info](https://www.radio-browser.info/#/) catalog of Internet radio stations by using the [project's api](https://de1.api.radio-browser.info/).

```php
use \AdinanCenci\RadioBrowser\RadioBrowser;

$browser          = new RadioBrowser();
$tag              = 'metal';
$orderBy          = 'name';
$descendingOrder  = true;

$stations         = $browser->getStationsByTag($tag, $orderBy, $descendingOrder);

print_r($stations);
```

<br><br>

## Examples

See examples of how to use in the "examples" directory.

- [Get tags](examples/a-get-tags.php)
- [Get states](examples/b-get-states.php)
- [Get stations by tag](examples/c-get-stations-by-tag.php)
- [Get stations by country](examples/d-get-stations-by-country.php)
- [Get stations by clicks](examples/e-get-stations-by-clicks.php)
- [Get stations by votes](examples/f-get-stations-by-votes.php)
- [Get stations in XML format](examples/g-get-stations-in-xml-format.php)
- [Get servers IPs](examples/y-get-servers-ips.php)

<br><br>

## Classes and instantiating

### Associative arrays / Objects

The methods of `\AdinanCenci\RadioBrowser\RadioBrowser` will return either associative arrays or stdObjects, depending on the `$associative` parameter informed to the constructor.

| Parameter    | Type         | Default                             | Description                                                                                                                                                                            |
| ------------ | ------------ | ----------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| $server      | string\|bool | https://de1.api.radio-browser.info/ | A Radio Browser server, it defaults to "'https://de1.api.radio-browser.info/", if false is informed the object will pick a random server, see the server section for more information. |
| $associative | bool         | true                                | If true, the methods will return associative arrays, stdObjects otherwise.                                                                                                             |

<br><br>

### Xml, json, csv etc...

If you need the data represented in a specific format, then use `\AdinanCenci\RadioBrowser\RarioBrowserApi`. The different methods will return data formatted as: json, xml, csv, m3u, pls, xspf and ttl.

| Parameter | Type         | Default                             | Description                                                                                                                                                                             |
| --------- | ------------ | ----------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| $server   | string\|bool | https://de1.api.radio-browser.info/ | A Radio Browser server, it defaults to "'https://de1.api.radio-browser.info/", if false is informed, the object will pick a random server, see the server section for more information. |
| $format   | string       | json                                | The format the data must be served, possible values: json, xml, csv, m3u, pls, xspf, ttl.                                                                                               |

<br><br>

## Radio Stations

### Get stations by UUID

The `::getStationsByUuid($uuids)` search stations with the specified ids.

| Parameter | Type          | Description                                                     |
| --------- | ------------- | --------------------------------------------------------------- |
| $uuids    | string\|array | A list of uuids, either an array, or an comma separated string. |

<br><br>

### Get stations by URL

The `::getStationsByUrl($url)`method search stations by their web page.

<br><br>

### Get stations by clicks

The `::getStationsByClicks($offset, $limit, $hideBroken)` method returns the most sintonized stations.

<br><br>

### Get stations by votes

The `::getStationsByVotes($offset, $limit, $hideBroken)` method returns the most voted stations.

<br><br>

### Get stations by recent clicks

The `::getStationsByRecentClicks($offset, $limit, $hideBroken)` method returns the currently most popular stations.

<br><br>

### Get stations last changed

The `::getStationsByLastChange($offset, $limit, $hideBroken)` method returns the stations last updated.

<br><br>

### Get older version of stations

The `::getStationOlderVersions($lastChangeUuid, $limit)` method returns old versions of stations from the last 30 days.

<br><br>

### Get broken stations

The `::getBrokenStations($offset, $limit)` method returns stations that did not pass the connection test.

<br><br>

## Get stations by ...

All the methods in this section share the following parameters:

| Parameter   | Type   | Default | Description                                                                                                                                                                            |
| ----------- | ------ | ------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| $order      | string | 'name'  | Possible values:<br />name, url, homepage, favicon, tags, country, state, language, votes, codec, bitrate, lastcheckok, lastchecktime, clicktimestamp, clickcount, clicktrend, random. |
| $reverse    | bool   | false   | false = Ascending order.<br />true = Descending order.                                                                                                                                 |
| $hideBroken | bool   | false   | Do not list stations that failed the connection test.                                                                                                                                  |
| $offset     | int    | 0       |                                                                                                                                                                                        |
| $limit      | int    | 100000  |                                                                                                                                                                                        |

<br><br>

### Just get all stations

The `::getStations($order, $reverse, $hideBroken, $offset, $limit)` method will return all stations.

<br><br>

### Get stations by name

The `::getStationsByName($name, $order, $reverse, $hideBroken, $offset, $limit)` method returns stations described with `$name`.

<br><br>

### Get stations by exact name

The `::getStationsByExactName($name, $order, $rev...)` method returns stations described with an exact match of `$name`.

<br><br>

### Get stations by codec

The `::getStationsByCodec($codec, $order, $rev...)` method returns stations described with `$codec`.

<br><br>

### Get stations by exact codec

The `::getStationsByExactCodec($codec, $order, $rev...)` method returns stations described with an exact match of `$codec`.

<br><br>

### Get stations by country

`::getStationsByCountry($country, $order, $rev...)`

<br><br>

### Get station by exact country

`::getStationsByExactCountry($country, $order, $rev...)`

<br><br>

### Get stations by state

`::getStationsByState($state, $order, $rev...)`

<br><br>

### Get stations by exact state

`::getStationsByExactState($state, $order, $rev...)`

<br><br>

### Get stations by language

`::getStationsByLanguage($language, $order, $rev...)` 

<br><br>

### Get stations by exact language

`::getStationsByExactLanguage($language, $order, $rev...)`.

<br><br>

### Get stations by tag

`::getStationsByTag($tag, $order, $rev...)`

<br><br>

### Get stations by exact tag

`::getStationsByExactTag($tag, $order, $rev...)`

<br><br>

## Search station

The `::searchStation($searchTerms)` method allow us to fine grain our search.
It receives a single associative array with the following keys available, all of which are optional:

| Key           | Type          | Default | Description                                                                                                                                                                            |
| ------------- | ------------- | ------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| name          | string        | null    |                                                                                                                                                                                        |
| nameExact     | bool          | false   |                                                                                                                                                                                        |
| country       | string        | null    |                                                                                                                                                                                        |
| countryExact  | bool          | false   |                                                                                                                                                                                        |
| countrycode   | string        | null    |                                                                                                                                                                                        |
| state         | string        | null    |                                                                                                                                                                                        |
| stateExact    | bool          | false   |                                                                                                                                                                                        |
| language      | string        | null    |                                                                                                                                                                                        |
| languageExact | bool          | false   |                                                                                                                                                                                        |
| tag           | string        | null    |                                                                                                                                                                                        |
| tagExact      | bool          | false   |                                                                                                                                                                                        |
| tagList       | string\|array | null    | A list of tags, either an array or a comma separated string.                                                                                                                           |
| codec         | string        | null    |                                                                                                                                                                                        |
| bitrateMin    | int           | 0       |                                                                                                                                                                                        |
| bitrateMax    | int           | 1000000 |                                                                                                                                                                                        |
| order         | string        | name    | Possible values:<br />name, url, homepage, favicon, tags, country, state, language, votes, codec, bitrate, lastcheckok, lastchecktime, clicktimestamp, clickcount, clicktrend, random. |
| reverse       | bool          | false   | false = Ascending order.<br />true = Descending order.                                                                                                                                 |
| offset        | int           | 0       |                                                                                                                                                                                        |
| limit         | int           | 100000  |                                                                                                                                                                                        |

<br><br>

## Get station checks

The `::getStationCheckResults($stationUuid, $lastCheckUuid, $seconds, $limit)` method returns a list of station check results. If a station UUID is provided, the whole history will be returned, otherwise a list of all last checks of all stations will be sent (without older check results).

| Parameter      | Type     | Default                                                        | Description                                                                   |
| -------------- | -------- | -------------------------------------------------------------- | ----------------------------------------------------------------------------- |
| $stationUiid   | 'string' |                                                                | Optional. If set, only list check result of the matching station.             |
| $lastCheckUuid | 'string' | If set, only list checks after the check with the given check. |                                                                               |
| $seconds       | integer  | 0                                                              | if > 0, it will only return history entries from the last `$seconds` seconds. |
| $limit         | integer  | 999999                                                         |                                                                               |

<br><br>

## Get station clicks

The `::getStationClicks($stationUuid, $lastCheckUuid, $seconds)` method returns a list of station clicks. If a station UUID is provided, only clicks of the station will be returned, otherwise a list of all clicks of all stations will be sent (chunksize 10000).

| Parameter      | Type     | Default                                                        | Description                                                                   |
| -------------- | -------- | -------------------------------------------------------------- | ----------------------------------------------------------------------------- |
| $stationUiid   | 'string' |                                                                | Optional. If set, only list check result of the matching station.             |
| $lastCheckUuid | 'string' | If set, only list checks after the check with the given check. |                                                                               |
| $seconds       | integer  | 0                                                              | if > 0, it will only return history entries from the last `$seconds` seconds. |

<br><br>

## Add station

The `::addStation($name, $url, $homePage, $favIcon, $countryCode, $state, $language, $tags, $geoLat, $geoLong)` method allow 
us to insert new stations.

It receives a single associative array with the following keys available, all of which are optional:

| Key          | Type   | Default | Description                                                                 |
| ------------ | ------ | ------- | --------------------------------------------------------------------------- |
| $name        | string |         | MANDATORY, the name of the radio station. Max 400 chars.                    |
| $rl          | string |         | MANDATORY, the URL of the station.                                          |
| $homePage    | string | null    | the homepage URL of the station.                                            |
| $favIcon     | string | null    | the URL of an image file (jpg or png).                                      |
| $countryCode | string | null    | The 2 letter countrycode of the country where the radio station is located. |
| $state       | string | null    | The name of the part of the country where the station is located.           |
| $language    | string | null    | The main language used in spoken text parts of the radio station.           |
| $tags        | string | null    | A list of tags separated by commas to describe the station.                 |
| $geoLat      | string | null    | The latitude of the stream location.                                        |
| $geoLong     | string | null    | The longitude of the stream location.                                       |

<br><br>

## Ranking

### Listeners

The `::clickStation($stationUuid)` method must be invoked every time a user starts playing a stream, this helps Radio Browser sort how popular each station is. **IMPORTANT**: Every call from the same IP address and for the same station only gets counted once per day.

### Voting

The `::voteStation($stationUuid)` method increases the vote count by one. **IMPORTANT**: it can only be called once every 10 minutes for the same radio stations, from the same IP.

<br><br>

## General Information

The methods bellow share the following parameters:

| Parameter   | Type   | Default | Description                                            |
| ----------- | ------ | ------- | ------------------------------------------------------ |
| $filter     | string |         | A string to be matched against.                        |
| $order      | string | name    | Possible values: name, stationcount.                   |

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.1.0] - 2022-03-05

### Changed
- Tidying up the documentation.

### Added
- Documentation for the methods `::getStationCheckResults()` and `::getStationClicks()`.
- Added the `$limit` parameter for the `::getStationCheckResults()` method.

### Fixed
 - Bug: without specifying the uuid, `::getStationCheckResults()` and `::getStationClicks()` would return empty results.

## [2.0.1] - 2021-11-01

### Changed

Added `guzzlehttp/guzzle` to the project, replacing my own `adinan-cenci/simple-request`.

## [2.0.0] - 2021-05-22

### Changed

The `RadioBrowser` class is now a wrapper for the new `RadioBrowserApi` class.
`RadioBrowserApi` is the general one, its methods return strings, `RadioBrowser` methods return arrays or objects depending on the parameters informed to the constructor.

### Added

- RadioBrowserApi class
- The read-only proprieties $server and $format.
- ::getStationsByClicks()
- ::getStationsByVotes()
- ::getStationsByRecentClicks()
- ::getStationsByLastChange()
- ::getStationOlderVersions()
- ::getBrokenStations()
- ::getServerStats()
- ::getServerMirrors()
- ::getServerConfig()
- ::getServerMetrics()
- ::addStation()
- added examples files

## [0.1.1] - 2021-01-28

### Fixed
- `::getStationsByExactTag($tag)` and `::getStationsByTag($tag)` no longer strip "#" from `$tag` before request.


src/RadioBrowserApi.php
<?php 
namespace AdinanCenci\RadioBrowser;

use \GuzzleHttp\Client;

class RadioBrowserApi 
{
    /** @var string $server The server URI */
    protected $server = null;

    /** @var string $format Possible values: json, xml, csv, m3u, pls, xspf, ttl */
    protected $format = 'json';

    public function __construct($server = 'https://de1.api.radio-browser.info/', $format = 'json') 
    {
        if ($server == false) {
            $server = self::pickAServer();
        }

        $this->server = rtrim($server, '/').'/';

        $this->format = $format;
    }

    public function __get($var) 
    {
        if (in_array($var, ['server', 'format'])) {
            return $this->{$var};
        }
        
        return null;
    }

    public function getCountries($filter = '', $order = 'name', $reverse = false, $hideBroken = false) 
    {
        return $this->standardRequest($this->server.$this->format.'/countries/'.$filter, $order, $reverse, $hideBroken);
    }

    public function getCountryCodes($filter = '', $order = 'name', $reverse = false, $hideBroken = false) 
    {
        return $this->standardRequest($this->server.$this->format.'/countrycodes/'.$filter, $order, $reverse, $hideBroken);
    }

    public function getCodecs($filter = '', $order = 'name', $reverse = false, $hideBroken = false) 
    {
        return $this->standardRequest($this->server.$this->format.'/codecs/'.$filter, $order, $reverse, $hideBroken);
    }

    public function getStates($filter = '', $country = null, $order = 'name', $reverse = false, $hideBroken = false) 
    {
        $url = $country ?
            $this->server.$this->format.'/states/'.$country.'/'.$filter : 
            $this->server.$this->format.'/states/'.$filter;

        $variables = [
            'order'      => $order, 
            'reverse'    => self::stringBoolean($reverse), 
            'hidebroken' => self::stringBoolean($hideBroken)
        ];
        
        return $this->fetchBody($url, $variables);
    }

    public function getLanguages($filter = '', $order = 'name', $reverse = false, $hideBroken = false) 
    {
        return $this->standardRequest($this->server.$this->format.'/languages/'.$filter, $order, $reverse, $hideBroken);
    }

    public function getTags($filter = '', $order = 'name', $reverse = false, $hideBroken = false) 
    {
        return $this->standardRequest($this->server.$this->format.'/tags/'.$filter, $order, $reverse, $hideBroken);
    }

    //------------------------------------

    public function getStationsByUuid($uuids) 
    {
        $parameters['uuids'] = is_array($uuids) ? implode(',', $uuids) : $uuids;

        $url = $this->server.$this->format.'/stations/byuuid';

        return $this->fetchBody($url, $parameters);
    }

    public function getStationsByUrl($url) 
    {
        $parameters['url'] = $url;

        $url = $this->server.$this->format.'/stations/byurl';

        return $this->fetchBody($url, $parameters);
    }

    public function getStationsByName($name, $order = 'name', $reverse = false, $hideBroken = false, $offset = 0, $limit = 100000) 
    {
        return $this->getStationsBy($this->server.$this->format.'/stations/byname/'.$name, $order, $reverse, $hideBroken, $offset, $limit);
    }

    public function getStationsByExactName($name, $order = 'name', $reverse = false, $hideBroken = false, $offset = 0, $limit = 100000) 
    {
        return $this->getStationsBy($this->server.$this->format.'/stations/bynameexact/'.$name, $order, $reverse, $hideBroken, $offset, $limit);
    }

    public function getStationsByCodec($codec, $order = 'name', $reverse = false, $hideBroken = false, $offset = 0, $limit = 100000) 
    {
        return $this->getStationsBy($this->server.$this->format.'/stations/bycodec/'.$codec, $order, $reverse, $hideBroken, $offset, $limit);
    }

    public function getStationsByExactCodec($codec, $order = 'name', $reverse = false, $hideBroken = false, $offset = 0, $limit = 100000) 
    {
        return $this->getStationsBy($this->server.$this->format.'/stations/bycodecexact/'.$codec, $order, $reverse, $hideBroken, $offset, $limit);
    }

    public function getStationsByCountry($country, $order = 'name', $reverse = false, $hideBroken = false, $offset = 0, $limit = 100000) 
    {
        return $this->getStationsBy($this->server.$this->format.'/stations/bycountry/'.$country, $order, $reverse, $hideBroken, $offset, $limit);
    }

    public function getStationsByExactCountry($country, $order = 'name', $reverse = false, $hideBroken = false, $offset = 0, $limit = 100000) 
    {
        return $this->getStationsBy($this->server.$this->format.'/stations/bycountryexact/'.$country, $order, $reverse, $hideBroken, $offset, $limit);
    }

    public function getStationsByState($state, $order = 'name', $reverse = false, $hideBroken = false, $offset = 0, $limit = 100000) 
    {
        return $this->getStationsBy($this->server.$this->format.'/stations/bystate/'.$state, $order, $reverse, $hideBroken, $offset, $limit);
    }

    public function getStationsByExactState($state, $order = 'name', $reverse = false, $hideBroken = false, $offset = 0, $limit = 100000) 
    {
        return $this->getStationsBy($this->server.$this->format.'/stations/bystateexact/'.$state, $order, $reverse, $hideBroken, $offset, $limit);
    }

    public function getStationsByLanguage($language, $order = 'name', $reverse = false, $hideBroken = false, $offset = 0, $limit = 100000) 
    {
        return $this->getStationsBy($this->server.$this->format.'/stations/bylanguage/'.$language, $order, $reverse, $hideBroken, $offset, $limit);
    }

    public function getStationsByExactLanguage($language, $order = 'name', $reverse = false, $hideBroken = false, $offset = 0, $limit = 100000) 
    {
        return $this->getStationsBy($this->server.$this->format.'/stations/bylanguageexact/'.$language, $order, $reverse, $hideBroken, $offset, $limit);
    }

    public function getStationsByTag($tag, $order = 'name', $reverse = false, $hideBroken = false, $offset = 0, $limit = 100000) 
    {
        $tag = rawurlencode( $tag );
        return $this->getStationsBy($this->server.$this->format.'/stations/bytag/'.$tag, $order, $reverse, $hideBroken, $offset, $limit);
    }

    public function getStationsByExactTag($tag, $order = 'name', $reverse = false, $hideBroken = false, $offset = 0, $limit = 100000) 
    {
        $tag = rawurlencode( $tag );
        return $this->getStationsBy($this->server.$this->format.'/stations/bytagexact/'.$tag, $order, $reverse, $hideBroken, $offset, $limit);
    }

    public function searchStation($searchTerms) 
    {
        $defaultParameters = [
            'name'          => null, 
            'nameExact'     => false, 
            'country'       => null, 
            'countryExact'  => false, 
            'countrycode'   => null, 
            'state'         => null, 
            'stateExact'    => false, 
            'language'      => null, 
            'languageExact' => false, 
            'tag'           => null, 
            'tagExact'      => false, 
            'tagList'       => null, 
            'codec'         => null, 
            'bitrateMin'    => 0, 
            'bitrateMax'    => 1000000, 
            'order'         => 'name', 
            'reverse'       => false, 
            'offset'        => 0, 
            'limit'         => 100000
        ];

        $parameters = array_merge($defaultParameters, $searchTerms);

        $parameters['nameExact']    = self::stringBoolean($parameters['nameExact']);
        $parameters['countryExact'] = self::stringBoolean($parameters['countryExact']);
        $parameters['stateExact']   = self::stringBoolean($parameters['stateExact']);
        $parameters['languageExact']= self::stringBoolean($parameters['languageExact']);
        $parameters['tagExact']     = self::stringBoolean($parameters['tagExact']);
        $parameters['reverse']      = self::stringBoolean($parameters['reverse']);

        if ($parameters['tag']) {
            $parameters['tag']      = is_array($parameters['tag']) ? implode(',', $parameters['tag']) : $parameters['tag'];
        }

        $url = $this->server.$this->format.'/stations/search';

        return $this->fetchBody($url, $parameters);
    }

    # all stations
    public function getStations($order = 'name', $reverse = false, $hideBroken = false, $offset = 0, $limit = 100000) 
    {
        return $this->getStationsBy($this->server.$this->format.'/stations', $order, $reverse, $hideBroken, $offset, $limit);
    }

    //------------------------------------

    public function getStationCheckResults($stationUuid = '', $lastCheckUuid = null, $seconds = 0) 
    {
        $url = $this->server.$this->format.'/checks/'.$stationUuid;

        $variables = [
            'lastcheckuuid' => $lastCheckUuid, 
            'seconds'       => $seconds
        ];

        return $this->fetchBody($url, $variables);
    }

    public function getStationClicks($stationUuid = '', $lastClickUuid = null, $seconds = 0) 
    {
        $url = $this->server.$this->format.'/clicks/'.$stationUuid;

        $variables = [
            'lastclickuuid' => $lastClickUuid, 
            'seconds'       => $seconds
        ];

        return $this->fetchBody($url, $variables);
    }

    public function clickStation($stationUuid) 
    {
        $url = $this->server.$this->format.'/url/'.$stationUuid;
        return $this->fetchBody($url);
    }

    public function voteStation($stationUuid) 
    {
        $url = $this->server.$this->format.'/vote/'.$stationUuid;
        return $this->fetchBody($url);
    }

    public function getStationsByClicks($offset = 0, $limit = 100000, $hideBroken = false) 
    {
        $url = $this->server.$this->format.'/stations/topclick';
        
        $variables = [
            'offset'     => $offset, 
            'limit'      => $limit, 
            'hidebroken' => self::stringBoolean($hideBroken)
        ];

        return $this->fetchBody($url, $variables);
    }

    public function getStationsByVotes($offset = 0, $limit = 100000, $hideBroken = false) 
    {
        $url = $this->server.$this->format.'/stations/topvote';
        
        $variables = [
            'offset'     => $offset, 
            'limit'      => $limit, 
            'hidebroken' => self::stringBoolean($hideBroken)
        ];

        return $this->fetchBody($url, $variables);
    }

    public function getStationsByRecentClicks($offset = 0, $limit = 100000, $hideBroken = false) 
    {
        $url = $this->server.$this->format.'/stations/lastclick';
        
        $variables = [
            'offset'     => $offset, 
            'limit'      => $limit, 
            'hidebroken' => self::stringBoolean($hideBroken)
        ];

        return $this->fetchBody($url, $variables);
    }

    public function getStationsByLastChange($offset = 0, $limit = 100000, $hideBroken = false) 
    {
        $url = $this->server.$this->format.'/stations/lastchange';
        
        $variables = [
            'offset'     => $offset, 
            'limit'      => $limit, 
            'hidebroken' => self::stringBoolean($hideBroken)
        ];

        return $this->fetchBody($url, $variables);
    }

    public function getStationOlderVersions($lastChangeUuid = '', $limit = 999999) 
    {
        $url = $this->server.$this->format.'/changed/'.$lastChangeUuid;
        
        $variables = [
            'limit' => $limit
        ];

        return $this->fetchBody($url, $variables);
    }

    public function getBrokenStations($offset = 0, $limit = 100000) 
    {
        $url = $this->server.$this->format.'/stations/broken';
        
        $variables = [
            'offset'     => $offset, 
            'limit'      => $limit
        ];

        return $this->fetchBody($url, $variables);
    }

    public function addStation($name, $url, $homePage = null, $favIcon = null, $countryCode = null, $state = null, $language = null, $tags = null, $geoLat = null, $geoLong = null) 
    {
        $url = $this->server.$this->format.'/add';

        $variables = [
            'name'        => $name, 
            'url'         => $url, 
            'homepage'    => $homePage, 
            'favicon'     => $favIcon, 
            'countrycode' => $countryCode, 
            'state'       => $state, 
            'language'    => $language, 
            'tags'        => $tags, 
            'geo_lat'     => $geoLat, 
            'geo_long'    => $geoLong
        ];

        return $this->fetchBody($url, $variables);
    }

    public function getServerStats() 
    {
        $url = $this->server.$this->format.'/stats';
        return $this->fetchBody($url);
    }

    public function getServerMirrors() 
    {
        $url = $this->server.$this->format.'/servers';
        return $this->fetchBody($url);
    }

    public function getServerConfig() 
    {
        $url = $this->server.$this->format.'/config';
        return $this->fetchBody($url);
    }

    public function getServerMetrics() 
    {
        $url = $this->server.'/metrics';
        return $this->fetchBody($url);
    }

    //------------------------------------

    // Most api calls support the exact same 3 parameters
    protected function standardRequest($url, $order = 'name', $reverse = false, $hideBroken = false) 
    {
        $variables = [
            'order'      => $order, 
            'reverse'    => self::stringBoolean($reverse), 
            'hidebroken' => self::stringBoolean($hideBroken)
        ];

        return $this->fetchBody($url, $variables);
    }    

    protected function getStationsBy($url, $order = 'name', $reverse = false, $hideBroken = false, $offset = 0, $limit = 100000) 
    {
        $variables = [
            'order'      => $order, 
            'reverse'    => self::stringBoolean($reverse), 
            'hidebroken' => self::stringBoolean($hideBroken), 
            'offset'     => $offset, 
            'limit'      => $limit
        ];

        return $this->fetchBody($url, $variables);
    }

    protected function fetch($url, $fields = array()) 
    {
        $client = new Client();

        $response = $client->request('GET', $url, ['query' => $fields]);
        if ($response->getStatusCode() != 200) {
            throw new \Exception('Error requesting "'.$url.'", code: '.$response->getStatusCode(), 1);
        }

        return $response;
    }

    protected function fetchBody($url, $fields = array()) 
    {
        $response = $this->fetch($url, $fields);
        return $response->getBody();
    }

    //------------------------------------

    public static function getDnsRecords() 
    {
        return dns_get_record('all.api.radio-browser.info', \DNS_A);
    }

    public static function getServerIps() 
    {
        $ips     = [];
        $records = self::getDnsRecords();

        foreach ($records as $r) {
            $ips[] = $r['ip'];
        }

        return $ips;
    }

    public static function getServers() 
    {
        $servers = [];

        foreach (self::getServerIps() as $ip) {
            $servers[] = gethostbyaddr($ip);
        }

        return $servers;
    }

    // pick a random server
    public static function pickAServer() 
    {
        $ips    = self::getServers();
        $count  = count($ips);
        $chosen = rand(0, $count - 1);
        return $ips[$chosen];
    }

    //------------------------------------

    protected static function stringBoolean($value) 
    {
        if ($value === 'false' || $value == false) {
            return 'false';
        }

        return 'true';
    }
}
